### ISSUE TITLE
use .subscribe({ fetchPolicy: 'cache-only' })

### ISSUE THREAD

**Issue by pospi** [Carbon-Farm-Network/app-carbon-farm-network#20]
In other work the easiest way of managing this has been via DOM events.

Unfortunately there is not a lot in the way of documentation on [managing cache updates in GraphQL mutations](https://www.apollographql.com/docs/react/caching/cache-interaction#example-updating-the-cache-after-a-mutation), but once understood it is quite a powerful optimisation measure.

1. Use any `mutation` callback's `update` option to [specify custom handling to manage the query cache](https://github.com/neighbour-hoods/timetracking-applet/blob/9fc537486d3c5bb301d32c71db3aee732dc1b82f/ui/src/components/valueflows/WorkInputManual/WorkInputManual.ts#L137)
    - note that unless complete results are sent to the cache in full, you will get errors from the GraphQL runtime and the values will not be inserted.
2. [Trigger a DOM event](https://github.com/neighbour-hoods/timetracking-applet/blob/9fc537486d3c5bb301d32c71db3aee732dc1b82f/ui/src/components/valueflows/WorkInputManual/WorkInputManual.ts#L214) in the component immediately after it has finished the `mutation()` call.
3. In a related (parent) component which needs to sync with the update from the (nested) child component, [listen for this event](https://github.com/neighbour-hoods/timetracking-applet/blob/9fc537486d3c5bb301d32c71db3aee732dc1b82f/ui/src/provider-app.ts#L46). In the handler, call `.subscribe({ fetchPolicy: 'cache-only' })` on any affected `query` instances managed by the component.
    - Note that you can expose such `ReadableQuery` objects (generated from `query()`) in child components so that a parent 'controller' component can manage linked updates across non-nested children (as per the linked example).
    - Note also that if the `fetchPolicy: 'cache-only'` is omitted, the data may will be re-queried from Holochain instead of the updated browser cache.

**Comment by pospi** [comment]
@LeosPrograms let me know how you go with this one once all other work on Facets & whatever user requirements are outstanding is completed. Not terribly urgent an issue but should result in more immediate updates throughout the application, and resolve any situations where screens are not updated as expected. Hopefully the code in the linked examples is enough to follow the principles along, even though it's a Lit app rather than a Svelte one (and all the event binding [conveniences](https://svelte.dev/tutorial/dom-events) of Svelte should make it much simpler to implement in this app).

**Comment by pospi** [comment]
Yep, that sounds like a culprit. If it's causing issues I will add it to the higher priority list.

We don't *have* to use the cache, really refetching from Holochain should be almost as quick. So if we are feeling pressured for time we could skip point 1 in the original issue and implement 3 without the `fetchPolicy: 'cache-only'` as that will fetch it from Holochain instead.

**Comment by pospi** [comment]
@LeosPrograms please feel free to move this issue to the "later..." milestone after the quick fix for the Offers page is completed, and change the title to something like "Leverage GraphQL cache for related component updates" since that will be the only remaining work thereafter.

**Comment by pospi** [comment]
I notice there are also some dirty workarounds using polling in some components. For example in `routes/+page.svelte`:

```ts
      setInterval(function(){
        fetchAgents()
      }, 20000)
```

These measures should be removed and instead the update coordinated as needed via DOM events as above.

**Comment by pospi** [comment]
I think that's fine for now? If you did not yet implement the `update` callback in mutations to refresh the cache (doesn't seem so?) then it wouldn't have any effect anyway. So simply calling `.subscribe()` without arguments is fine to trigger an update in a parent component; and the optimisation is only a minor one since the data will be re-pulled from Holochain's caches anyway.

Will leave this issue open, but set it to a low priority cleanup milestone.

**Comment by fosterlynn** [comment]
Should we close this?

**Comment by LeosPrograms** [comment]
Yes
